tracker/view/login.py

```diff
@@ -8,39 +8,105 @@
+        return redirect(url_for('tracker.list_user'))
+    else:
+        if current_user.is_authenticated:
+            return redirect(url_for('tracker.index'))
 
-    form = LoginForm()
-    if not form.validate_on_submit():
-        status_code = Unauthorized.code if form.is_submitted() else 200
-        return render_template('login.html',
-                               title='Login',
-                               form=form,
-                               User=User,
-                               password_length={'min': TRACKER_PASSWORD_LENGTH_MIN,
-                                                'max': TRACKER_PASSWORD_LENGTH_MAX}), status_code
-
-    user = user_assign_new_token(form.user)
-    user.is_authenticated = True
-    login_user(user)
-    return redirect(url_for('tracker.index'))
+        form = LoginForm()
+        if not form.validate_on_submit():
+            status_code = Unauthorized.code if form.is_submitted() else 200
+            return render_template('login.html',
+                                title='Login',
+                                form=form,
+                                User=User,
+                                password_length={'min': TRACKER_PASSWORD_LENGTH_MIN,
+                                                 'max': TRACKER_PASSWORD_LENGTH_MAX}), status_code
+
+        user = user_assign_new_token(form.user)
+        user.is_authenticated = True
+        login_user(user)
+        return redirect(url_for('tracker.index'))
 
 
 @tracker.route('/logout', methods=['GET', 'POST'])
 def logout():
+    # TODO clear SSO session
     if not current_user.is_authenticated:
         return redirect(url_for('tracker.index'))
 
     user_invalidate(current_user)
     logout_user()
     return redirect(url_for('tracker.index'))
+
+
+@tracker.route('/sso-auth')
+def sso_auth():
+    from tracker import db
+
+    # login the user here, create a session and set role
+    token = oauth.idp.authorize_access_token()
+    parsed_token = oauth.idp.parse_id_token(token)
+    # check if user can be matched against local db of users
+    user = db.get(User, email=parsed_token.get('email'))
+    user_groups = parsed_token.get('groups')
+    user_groups_present = user_groups != None
+
+    if not user_groups_present or len(user_groups) == 0:
+        print("SSO error: a user authenticated without any valid groups")
+
+    current_maximum_role = condense_user_groups_to_role(user_groups)
+
+    if user:
+        if user.role != current_maximum_role:
+            user.role = current_maximum_role
+            db.session.add(user)
+            db.session.commit()
+
+        user = user_assign_new_token(user)
+        user.is_authenticated = True
+        login_user(user)
+    else:
+        # user does not exist in local db
+        # need to create him to leverage existing user access controls
+        from tracker.user import random_string, hash_password
+
+        user = User()
+        user.name = ''
```

Comment:
username should be required and fetched from the oidc token, I think the field should be `preferred_username`.

Reply:
I used that before but couldn't get it to work locally. I put it back in place now with an empty fallback value. I hope you can get it to work with the Arch setup